otodom/task_1/mk/otodom.py

```diff
@@ -0,0 +1,165 @@
+            req = requests.get(url, headers=HEADERS)
+        except requests.exceptions.RequestException:
+            print(f"Failed to retrieve the webpage. Status code: {req.status_code}")
+            return None
+        return BeautifulSoup(req.text, "html.parser")
+
+    def getLinks(self, url):
+        bs = self.getPage(url)
+        if bs is not None:
+            promoted_listings = bs.find(
+                "div", {"data-cy": "search.listing.promoted"}
+            ).find_all("a", {"class": "css-1hfdwlm e1dfeild2"})
+            promoted = []
+            for link in promoted_listings:
+                if "href" in link.attrs:
+                    promoted.append(link.attrs["href"])
+
+            organic_listings = bs.find(
+                "div", {"data-cy": "search.listing.organic"}
+            ).find_all("a", {"class": "css-1hfdwlm e1dfeild2"})
+            organic = []
+            for link in organic_listings:
+                if "href" in link.attrs:
+                    organic.append(link.attrs["href"])
+
+            return (promoted, organic)
+        return None
+
+    def getListing(self, pageUrl, promoted):
+        url = "http://www.otodom.pl{}".format(pageUrl)
+        bs = self.getPage(url)
+        if bs is not None:
+            listing = dict()
+            listing["url"] = url
+
+            id_match = re.search(
+                "[0-9]+$", bs.find("meta", {"name": "description"}).get("content", "")
+            )
+            listing["otodom_id"] = id_match.group() if id_match else ""
+
+            listing["title"] = bs.find("h1", {"class": "css-1wnihf5 efcnut38"}).text
+
+            localization = dict()
+            l10n = bs.find("a", {"class": "e1w8sadu0 css-1helwne exgq9l20"}).text.split(
+                ","
+            )
+
+            localization["province"] = l10n[-1] if len(l10n) >= 4 else ""
+            localization["city"] = l10n[-2] if len(l10n) >= 4 else l10n[-1]
+            localization["district"] = l10n[1] if len(l10n) >= 4 else ""
+            localization["street"] = l10n[0] if len(l10n) >= 4 else l10n[0]
+
+            listing["localization"] = localization
+
+            listing["promoted"] = promoted
+            number = bs.find("strong", {"class": "css-t3wmkv e1l1avn10"}).text.replace(
+                ",", "."
+            )
+            listing["price"] = int(float(re.sub(r"[^.0-9]", "", number)))
+            number = bs.find(
+                "div", {"data-testid": "table-value-rooms_num"}
+            ).text.replace(",", ".")
+            listing["rooms"] = int(float(re.sub(r"[^.0-9]", "", number)))
+            number = bs.find("div", {"data-testid": "table-value-area"}).text.replace(
+                ",", "."
+            )
+            listing["area"] = int(float(re.sub(r"[^.0-9]", "", number)))
+            listing["estate_agency"] = bs.find(
+                "div", {"data-testid": "table-value-advertiser_type"}
+            ).text
+
+            return listing
```

Comment:
btw all of this data should be possible to be scraped from the main page. there shouldn't be a need to additionaly make a request to the listings url
<img width="659" alt="image" src="https://github.com/Kolo-Naukowe-Data-Science-PW/Data-Scraping-Tasks-23-24/assets/111927572/4caa58b0-7f32-4572-b217-a32647b405bf">